janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -17,81 +17,236 @@
+            }
+            offsets[idx] = subQueryCache.size();
+        } else {
+            QueryProfiler profiler = subQuery.getProfiler();
+            QueryProfiler.startProfile(profiler, subQuery);
+            // TODO: leverage the scrolling capability of indexing backends ratherthan throw away old results
+            indexSerializer.query(subQuery, tx).skip(offsets[idx]).forEachOrdered(result -> {
+                offsets[idx]++;
+                subResultToQueryMap.computeIfAbsent(result, k -> new ArrayList<>()).add(idx);
+            });
             profiler.stopTimer();
+            profiler.setResultSize(offsets[idx]);
+            // TODO: should we put it into cache?
+        }
+
+    }
+
+    private void intersectResults(Map<Object, List<Integer>> subResultToQueryMap,
+                                  List<Object> results, final int limit, final int numberOfQueries) {
+        for (Iterator<Map.Entry<Object, List<Integer>>> it = subResultToQueryMap.entrySet().iterator();
+             it.hasNext(); ) {
+            Map.Entry<Object, List<Integer>> entry = it.next();
+            if (entry.getValue().size() == numberOfQueries) {
+                // this particular result satisfies every index query
+                results.add(entry.getKey());
+                if (results.size() >= limit) break;
+                it.remove();
+            }
+        }
+    }
+
+    /**
+     * Compute score for each query. Lower score means the query is more selective and more likely to
+     * be the bottleneck. Exhausted queries always have 0 score. Unless more factors are taken into consideration,
+     * at the moment it does not make sense to compare queries if we only have two.
+     */
+    private double[] computeScores(final int resultsExhaustedCount, boolean[] resultsExhausted,
+                                   int[] offsets, final int numberOfQueries, final int numberOfResults,
+                                   final int limit, Map<Object, List<Integer>> subResultToQueryMap) {
+        double[] scores = new double[numberOfQueries];
+        // TODO: take more factors into consideration, e.g. query latency
+        if (resultsExhaustedCount < numberOfQueries && numberOfResults < limit && numberOfQueries > 2) {
+            // A query whose results have many intersections with other queries is less likely to be selective.
+            // A query with few intersections with other queries is more valuable in the sense that it contains
+            // less redundant information, and thus is more useful.
+            // In other words, if multiple index queries have huge overlaps with each other except one query,
+            // then that query is more likely to be the bottleneck.
+            for (List<Integer> queryNoList : subResultToQueryMap.values()) {
+                for (int idx : queryNoList) {
+                    if (!resultsExhausted[idx]) scores[idx] += Math.log(queryNoList.size());
```

Comment:
I like this approach! Did you come up with it yourself or is there any scientific research that suggests this as a measure of selectivity?

Reply:
I came up with this by myself based on intuition. My thoughts might be wrong though.

My basic assumption is that no index is perfect; there are indexes that more or less store some sort of 'redundant' information. In the extreme case, some indexes might return exactly the same results.

Reply:
In fact, now I doubt if we should do such selection at all since we can possibly execute all backend queries in parallel (I don't understand why the original author did not do this in the first place).

Reply:
Parallelization sounds like a good idea here. Would you like to address it in this PR or rather leave it as a TODO?

Reply:
Not sure, I would like to hear more voice on this. A possible reason for not using parallelization is due to the additional overhead, especially when the overall read load is very heavy, even though latency for a single read request can be reduced drastically by parallelization.